app/src/androidTest/java/ch/epfl/polychef/HomePageUserTest.java

```diff
@@ -0,0 +1,260 @@
+
+        when(mockDatabase.getReference("users/" + mockUserKey)).thenReturn(mockNewUserRef);
+
+        assertSendingBackCorrectUser(mockNewUserRef, new User(mockUserEmail, mockUserName));
+    }
+
+    @Test
+    public void oldUserTest(){
+
+        onDataChangeCallBack();
+
+        when(mockOnDataChangeSnapshot.getChildrenCount()).thenReturn((long) 1);
+
+        //OnDataChange mock requirements
+        DataSnapshot mockSnapshotChild = mock(DataSnapshot.class);
+
+        List<DataSnapshot> children = new ArrayList<>(1);
+        children.add(mockSnapshotChild);
+        when(mockOnDataChangeSnapshot.getChildren()).thenReturn(children);
+
+        //oldUser method mock requirements
+        User mockUser = mockUser();
+
+        when(mockSnapshotChild.exists()).thenReturn(true);
+        when(mockSnapshotChild.getValue(User.class)).thenReturn(mockUser);
+        when(mockSnapshotChild.getKey()).thenReturn(mockUserKey);
+
+        //updateUser method mock requirements
+        DatabaseReference mockOldUserRef = mock(DatabaseReference.class);
+
+        when(mockDatabase.getReference("users/" + mockUserKey)).thenReturn(mockOldUserRef);
+
+        assertSendingBackCorrectUser(mockUsersRef, mockUser);
+    }
+
+    public void assertOnDataChangeThrowsException(Exception exception){
+        doAnswer((call) -> {
+            ValueEventListener listener =  call.getArgument(0);
+
+            assertThrows(exception.getClass(), () -> listener.onDataChange(mockOnDataChangeSnapshot));
+
+            return null;
+        }).when(mockEqualToEmail).addListenerForSingleValueEvent(any(ValueEventListener.class));
+    }
+
+    @Test
+    public void throwsExceptionWhenMultipleUsersExist() {
+
+        when(mockOnDataChangeSnapshot.getChildrenCount()).thenReturn((long) 2);
+
+        assertOnDataChangeThrowsException(new IllegalStateException());
+    }
+
+    @Test
+    public void testJsonErrorThrowsException(){
+
+        when(mockOnDataChangeSnapshot.getChildrenCount()).thenReturn((long) 1);
+
+        assertOnDataChangeThrowsException(new IllegalArgumentException());
+
+        //OnDataChange mock requirements
+        DataSnapshot mockSnapshotChild = mock(DataSnapshot.class);
+
+        List<DataSnapshot> children = new ArrayList<>(1);
+        children.add(mockSnapshotChild);
+        when(mockOnDataChangeSnapshot.getChildren()).thenReturn(children);
+
+        //oldUser method mock requirements
+        when(mockSnapshotChild.exists()).thenReturn(false);
+    }
+
+    @Test
+    public void throwsExceptionWhenQueryCancelled() {
+        doAnswer((call) -> {
+            ValueEventListener listener =  call.getArgument(0);
+
+            assertThrows(IllegalArgumentException.class, () -> listener.onCancelled(DatabaseError.fromException(new Exception())));
+
+            return null;
+        }).when(mockEqualToEmail).addListenerForSingleValueEvent(any(ValueEventListener.class));
+    }
+}
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
+
```

Comment:
No need to have so much blank lines at the end of the file

Reply:
Yep, my bad. Should have enabled "scroll past end" instead